### PR TITLE
secp256k1: Expose Jacobian point equivalency func.

### PR DESCRIPTION
**Pull request by davecgh**
This exposes a new function on the `JacobianPoint` type named `EquivalentNonConst` which efficiently determines if two Jacobian points represent the same affine point without actually converting the points to affine.

This provides a significant speedup versus first converting to affine for use cases that need the functionality.  One example where it is useful is adaptor signatures.

It includes comprehensive tests for edge conditions as well as ongoing randomized testing.

Finally, it also includes some preliminary commits to make the benchmark structure more consistent and introduce a new benchmark for testing Jacobian point equivalency.

The following benchmark shows a before and after comparison of checking Jacobian point equivalency with the new method versus the affine conversion approach:

```
name                       old time/op   new time/op  delta
--------------------------------------------------------------------------------
JacobianPointEquivalency   17.2µs ± 2%   0.5µs ± 1%   -97.24%  (p=0.000 n=10+10)
```

### CODE REVIEW

**Review by martonp — COMMENTED** [review]
LGTM. NBD, but should it be called `EquivalentNonConst`?

**Review by martonp — APPROVED** [review]

**Comment by davecgh** [comment]
> LGTM. NBD, but should it be called `EquivalentNonConst`?

Good point.  That would be less likely to potentially cause confusion of exact equality versus equivalent points.  Updated.

File: dcrec/secp256k1/curve_test.go [hunk 1]
```diff
@@ -56,6 +101,241 @@ func jacobianPointFromHex(x, y, z string) JacobianPoint {
...
+			continue
+		}
+
+		// Convert the points to affine and ensure they have the expected
+		// equivalency as well.
+		got := isSameAffinePoint(&p1, &p2)
+		if got != test.want {
+			t.Errorf("%s: mismatched expected test equivalency -- got %v, "+
+				"want %v", test.name, got, test.want)
+			continue
```
Comment (jrick) [inline dcrec/secp256k1/curve_test.go:287]: we're obviously not hitting these but this continue (and the one below) aren't needed, and it might be "interesting" to see a failed test where only one them didn't match expected (instead of this one causing the public method to not be checked)

**Review by jrick — APPROVED** [review]
ok with trivial suggestion

File: dcrec/secp256k1/curve_test.go (hunk 1 above)
Comment (davecgh) [reply dcrec/secp256k1/curve_test.go:287]: Updated.

File: dcrec/secp256k1/curve_test.go [hunk 2]
```diff
@@ -45,6 +46,50 @@ func isValidJacobianPoint(point *JacobianPoint) bool {
...
+	for randZ.IsZero() {
+		randZ = randFieldVal(t, rng)
+	}
+	var pt JacobianPoint
+	pubKey.AsJacobian(&pt)
+	pt.Rescale(randZ)
+
+	// Sanity check the result.
+	if !isValidJacobianPoint(&pt) {
+		t.Fatal("generatd random Jacobian point is not on the curve")
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve_test.go]: Fixed.

File: dcrec/secp256k1/curve_test.go [hunk 3]
```diff
@@ -68,6 +113,229 @@ func isSameAffinePoint(p1, p2 *JacobianPoint) bool {
...
 }
 
+// TestEquivalentJacobian ensures determining if two Jacobian points represent
+// the same affine point via [JacobianPoint.EquivalentNonConst] works as
+// intended for some edge cases and known values.  It also verifies in affine
+// coordinates as well.
+func TestEquivalentJacobian(t *testing.T) {
+	tests := []struct {
+		name       string // test description
+		x1, y1, z1 string // hex encoded coordinates of first point to add
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve_test.go]: Changed.

File: dcrec/secp256k1/curve_test.go [hunk 4]
```diff
@@ -68,6 +113,229 @@ func isSameAffinePoint(p1, p2 *JacobianPoint) bool {
...
+		x2, y2, z2 string // hex encoded coordinates of second point to add
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve_test.go]: Changed.

File: dcrec/secp256k1/curve_test.go (hunk 1 above)
Comment (matthawkins90) [reply dcrec/secp256k1/curve_test.go:287]: It doesn't look like anything changed.

Comment (davecgh) [reply dcrec/secp256k1/curve_test.go]: Thanks.  Forgot to commit it.  Now it's done.

**Review by jrick — APPROVED** [review]

File: dcrec/secp256k1/curve.go [hunk 5]
```diff
@@ -149,6 +149,18 @@ func (p *JacobianPoint) ToAffine() {
 	p.Y.Normalize()
 }
 
+// EquivalentNonConst returns whether or not two Jacobian points represent the
+// same affine point in *non-constant* time.
+func (p *JacobianPoint) EquivalentNonConst(other *JacobianPoint) bool {
+	// Use the group law that a point minus itself is the point at infinity to
```
Comment (davecgh) [inline dcrec/secp256k1/curve.go:155]: While this statement is true, it might not be obvious for those who aren't super familiar with group laws and their implications, so I wonder if rather than just asserting the implication this would be clearer as:

```Go
       // Since the point at infinity is the identity element for the group, note  
       // that P = P + ∞ trivially implies that P - P = ∞.
       //
       // Use that fact to determine if the points represent the same affine point.
```

Comment (jrick) [reply dcrec/secp256k1/curve.go:155]: it read fine as is before, but this is ok too.

Comment (davecgh) [reply dcrec/secp256k1/curve.go]: Updated.
